fuel/streams.py

```diff
@@ -37,11 +37,19 @@ class AbstractDataStream(object):
     sources : tuple of strings
         The names of the data sources returned by this data stream, as
         given by the dataset.
+    produces_examples : bool
+        Whether this data stream produces examples (as opposed to batches
+        of examples).
 
     """
     def __init__(self, iteration_scheme=None, axis_labels=None):
         self.iteration_scheme = iteration_scheme
         self.axis_labels = axis_labels
+        if self.iteration_scheme:
+            self.produces_examples = iteration_scheme.requests_examples
+        else:
+            # TODO: is this a valid assumption?
```

Comment:
See 2970001 for an overview of what changes it entails.

The biggest change from the users' point of view is that they'll have to get used to call `IterableDataset(...).get_example_stream()` instead of `DataStream(IterableDataset(...))`, because the latter would force them to add an additional `stream.produces_example = True` statement afterwards. I think it's a good trade-off.